src/backend/src/main/kotlin/au/com/cding21/third_party/banks/INGClientImpl.kt

```diff
@@ -0,0 +1,95 @@
+package au.com.cding21.third_party.banks
+
+import au.com.cding21.third_party.banks.allocators.SynchronousAllocator
+import au.com.cding21.third_party.banks.types.Account
+import au.com.cding21.third_party.banks.types.Transaction
+import au.com.cding21.third_party.banks.util.*
+import com.microsoft.playwright.BrowserContext
+import com.microsoft.playwright.Page
+import com.microsoft.playwright.TimeoutError
+import com.microsoft.playwright.options.LoadState
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.jsonPrimitive
+import java.util.*
+import javax.security.auth.login.CredentialException
+
+/**
+ * Stateless ING API client to access ING account data
+ */
+class INGClientImpl(
+    private val allocator: SynchronousAllocator,
+    private val username: String,
+    private val password: List<Int>
+) : BankClient {
+    companion object {
+        val IMAGE_REFERENCES: List<ImageCodec> = listOf(
```

Comment:
This is not ideal at scale. Ideally images can be differentiated without rendering it into a raster image and doing pixel comparisons.

`/PinpadImages` returns a RSA public key (used to encrypt scrambled keypad positions) and an unknown `secret` field that isn't used except being sent back to the server upon login. Since it is unlikely that plaintext information is persisted within the images themselves (e.g no hidden chunks), either

- State info is encrypted with `secret` and hidden within `IDAT` in the PNG
- Images are truly randomised and exists as states in the backend to decode positions, in which case there isn't anything to do except a more optimised way to parse image and perform OCR

Reply:
@cding21 Lol bit of context (and likely irrelevant for you anyways since jian is the only person i know who unironically uses ing), ING doesn't use passwords for logins and instead uses a 4-digit pin code that you have to click on their pin pad.

Problem is, they scramble the pin positions and communicate with their BE using scrambled position index. So for example a `[0, 3, 0, 1]` can mean vastly different things since index 0 can be any digit (since its all randomised). This would normally be fine (given that I'm using webdrivers), the problem however is that their pin pad images are actually randomised as well. There is no meta information in the DOM as well (i.e additional attrs that might indicate which image is which key code etc.).

The images themselves are fairly standard, with a header chunk (which seems to be consistent across all images), followed by an `IDAT` chunk (which contains randomised data) and `IEND` (which is also consistent), with no additional hidden chunks or data. This leads back to my 2 guesses as to how ING actually differentiate images in the BE.

The actual request to get images also contain a RSA public key, which is used to encrypt the pin indexes in transit, and a `secret` field that I'm not too sure what it does, doesn't seem to be used in the FE.

Reply:
<img src="https://media.tenor.com/9glrQrHmKR8AAAAM/hog-rider-coc.gif">

Reply:
Jokes aside I did take a look at the ING login page (interesting that they use a pin pad like that), agreed that the most reliable way would be to using your image matching method (don't mind the computationally expensiveness). 

What I meant in the previous comment was the encryption of the secret itself, would be interested if we could pry into it but I don't we'll ever be able to IF they use RSA without the private key. So yeah will have to rate-limit ING much harder than ANZ to limit resource usage.

My next thought is how to integrate all this to sync with the existing transaction endpoint. My current draft is to upload/sync this on a user request basis (will need a sync state to be persisted somewhere so the sync isn't accidentally abused, ie. "up-to-date, syncing & out-of-date" states) but happy to hear thoughts on this.

Reply:
Mmmm yeah, my original thought was to copy these bank transactions over to my own DB but the more thought I give it the more pain points arise. Will take this offline now, too long to discuss here...